cores/esp8266/core_esp8266_waveform.cpp

```diff
@@ -50,34 +50,27 @@ extern "C" {
 #define SetGPIO(a) do { GPOS = a; } while (0)
 #define ClearGPIO(a) do { GPOC = a; } while (0)
 
+enum class ExpiryState : uint32_t {OFF = 0, ON = 1, UPDATE = 2}; // for UPDATE, the NMI computes the exact expiry cycle and transitions to ON
+
 // Waveform generator can create tones, PWM, and servos
 typedef struct {
-  uint32_t nextServiceCycle;   // ESP cycle timer when a transition required
-  uint32_t expiryCycle;        // For time-limited waveform, the cycle when this waveform must stop
-  uint32_t nextTimeHighCycles; // Copy over low->high to keep smooth waveform
-  uint32_t nextTimeLowCycles;  // Copy over high->low to keep smooth waveform
+  uint32_t nextServiceCycle;           // ESP cycle timer when a transition required
+  volatile int32_t nextTimeHighCycles; // Copy over low->high to keep smooth waveform
```

Comment:
What's the logic of going `volatile` on these nextTimeXXX vars?  I think I only see it set in a single routine, so there's no chance of a mixup.  The volatile causes gcc to emit lots of `memw` which adds code size and stalls the pipeline slowing things down a bit...

Reply:
The reason is to prevent the compiler from reordering instructions. That's good the for lock-free state transfer to the IRQ. (you know, I usually go all-out and use std::atomic ;-) )